akka-persistence-typed-tests/src/test/scala/akka/persistence/typed/state/scaladsl/DurableStateBehaviorSpec.scala

```diff
@@ -0,0 +1,123 @@
+import akka.persistence.typed.PersistenceId
+import akka.serialization.jackson.CborSerializable
+
+object DurableStateBehaviorSpec {
+
+  def conf: Config = PersistenceTestKitDurableStateStorePlugin.config.withFallback(ConfigFactory.parseString(s"""
+    akka.loglevel = INFO
+    """))
+
+  sealed trait Command extends CborSerializable
+  final case class IncrementWithConfirmation(replyTo: ActorRef[Done]) extends Command
+  final case class GetValue(replyTo: ActorRef[State]) extends Command
+  final case class DeleteWithConfirmation(replyTo: ActorRef[Done]) extends Command
+  case class IncrementBy(by: Int) extends Command
+
+  final case class State(value: Int) extends CborSerializable
+
+  def counter(persistenceId: PersistenceId): DurableStateBehavior[Command, State] = {
+    DurableStateBehavior(
+      persistenceId,
+      emptyState = State(0),
+      commandHandler = (state, command) =>
+        command match {
+
+          case IncrementBy(by) =>
+            Effect.persist(state.copy(value = state.value + by))
+
+          case IncrementWithConfirmation(replyTo) =>
+            Effect.persist(state.copy(value = state.value + 1)).thenRun(_ => replyTo ! Done)
+
+          case GetValue(replyTo) =>
+            replyTo ! state
+            Effect.none
+
+          case DeleteWithConfirmation(replyTo) =>
+            Effect.delete[State]().thenRun(_ => replyTo ! Done)
+        })
+  }
+}
+
+class DurableStateBehaviorSpec
+    extends ScalaTestWithActorTestKit(NullEmptyStateSpec.conf)
+    with AnyWordSpecLike
+    with LogCapturing {
+  import DurableStateBehaviorSpec._
+
+  implicit val testSettings: TestKitSettings = TestKitSettings(system)
+
+  val pidCounter = new AtomicInteger(0)
+  private def nextPid(): PersistenceId = PersistenceId.ofUniqueId(s"c${pidCounter.incrementAndGet()})")
+
+  "A DurableStateBehavior actor" must {
+    "persist and update state" in {
+      val c = spawn(counter(nextPid()))
+      val updateProbe = TestProbe[Done]()
+      val queryProbe = TestProbe[State]()
+
+      c ! IncrementWithConfirmation(updateProbe.ref)
+      updateProbe.expectMessage(Done)
+
+      c ! GetValue(queryProbe.ref)
+      queryProbe.expectMessage(State(1))
+
+      c ! IncrementBy(5)
+      c ! IncrementWithConfirmation(updateProbe.ref)
+      updateProbe.expectMessage(Done)
+
+      c ! GetValue(queryProbe.ref)
+      queryProbe.expectMessage(State(7))
+    }
+
+    "delete state" in {
+      val c = spawn(counter(nextPid()))
+      val updateProbe = TestProbe[Done]()
+      c ! IncrementWithConfirmation(updateProbe.ref)
+      updateProbe.expectMessage(Done)
+
+      val deleteProbe = TestProbe[Done]()
+      c ! DeleteWithConfirmation(deleteProbe.ref)
+      deleteProbe.expectMessage(Done)
+
+      val queryProbe = TestProbe[State]()
+      c ! GetValue(queryProbe.ref)
+      queryProbe.expectMessage(State(0))
+    }
+
+    "handle commands sequentially" in {
```

Comment:
This was what I tried to reproduce the problem, but it is actually not visible from this test because we exposed the new state immediately so even if GetValue arrived too early it returned the "right" value.

I could confirm the bug with some strategic sleep and println.